lib/src/repo.rs

```diff
@@ -1664,34 +1665,38 @@ impl MutableRepo {
     }
 
     fn merge_view(&mut self, base: &View, other: &View) -> BackendResult<()> {
-        // TODO: Use `diff_named_commit_ids` to simplify this.
-        // Merge working-copy commits. If there's a conflict, we keep the self side.
-        for (workspace_id, base_wc_commit) in base.wc_commit_ids() {
-            let self_wc_commit = self.view().get_wc_commit_id(workspace_id);
-            let other_wc_commit = other.get_wc_commit_id(workspace_id);
-            if other_wc_commit == Some(base_wc_commit) || other_wc_commit == self_wc_commit {
-                // The other side didn't change or both sides changed in the
-                // same way.
-            } else if let Some(other_wc_commit) = other_wc_commit {
-                if self_wc_commit == Some(base_wc_commit) {
-                    self.view_mut()
-                        .set_wc_commit(workspace_id.clone(), other_wc_commit.clone());
-                }
+        let mut workspace_updates = vec![];
+        for (workspace_id, (my_commit, other_commit)) in
+            diff_named_commit_ids(self.view().wc_commit_ids(), other.wc_commit_ids())
+        {
+            // We never have `my_commit == other_commit` as
+            // `diff_named_commit_ids` omits such pairs. This is fine; we want
+            // to keep the `self` side unchanged in such cases.
+            let base_wc_commit = base.get_wc_commit_id(workspace_id);
+            let changed_commit = if my_commit == base_wc_commit {
+                other_commit
+            } else if other_commit == base_wc_commit {
+                // Keep the `self` side unchanged. Returning `my_commit` here
+                // would look more symmetric, but would have the same outcome.
+                continue;
+            } else if other_commit.is_none() {
+                // The other side removed the workspace and it moved on our
+                // side. We want to remove it. TODO(ilyagr): I'm not sure why.
```

Comment:
Updating a workspace to a different commit doesn't impact the commit graph or bookmarks so I don't think we need to preserve any information about that.

FWIW, what probably happened was that the user ran `jj workspace forget` in one workspace and something like `jj rebase` in another workspace. I don't think the user wants to get the workspace back because they ran that rebase at the same time.